Holiday api 스크립트 작성

cafitac opened the pull request:
api.py로 api 받아오는 스크립트 추가
Holiday 모델 추가
Holiday 모델 admin에 등록
local_settings의 API_KEY 값 변경

Commit:
fcm-django 설치

ADD: fcm-django (install, requirements.txt) 설치
ADD: settings.py INSTALLED_APPS에 추가

hanseongd commented on scheduler/schedule/api.py:
```diff
@@ -0,0 +1,59 @@
+# 특일 정보 api 키
+api_key = API_KEY
+
+# 국경일
```
API 주소 다 똑같은데 최대한 합쳐보기

hanseongd commented on scheduler/schedule/api.py:
```diff
@@ -0,0 +1,59 @@
+    if not Holiday.objects.filter(date=date, name=name):
+        Holiday.objects.create(name=name, is_holiday=is_holiday, date=date)
+
+        print("save")
```
print 제거

hanseongd commented on scheduler/schedule/api.py:
```diff
@@ -0,0 +1,59 @@
+    for year in range(2015, 2021):
+        year = str(year)
+        for month in range(1, 13):
+            if month // 10 < 1:
```
zfill 명령어 알아보고 바꾸기

hanseongd commented on scheduler/schedule/api.py:
```diff
@@ -0,0 +1,59 @@
+                is_holiday = item.find('isHoliday').text
+                date = item.find('locdate').text
+
+                if is_holiday == 'Y':
```
is_holiday = True if is_holiday == 'Y' else False

cafitac replied:
좋은 문법 알고 갑니당....

hanseongd commented on scheduler/schedule/api.py:
```diff
@@ -0,0 +1,59 @@
+        for month in range(1, 13):
+            if month // 10 < 1:
+                month = "0" + str(month)
+            response = requests.get(url .format(year=year, month=month, key=api_key))
```
url .format 띄어쓰기 수정

cafitac replied:
이 부분은 가독성 때문에 일부로 띄어쓰기를 해줬던건데 한 번만 다시 확인해주세욥
띄어쓰기를 없애야되면 다시 commit 해드리겠습니당

hanseongd replied:
이런 경우엔 가독성을 위해서 띄어쓰지 않아

hanseongd commented on scheduler/schedule/api.py:
```diff
@@ -0,0 +1,59 @@
+                else:
+                    is_holiday = False
+
+                date = date[:4] + "-" + date[4:6] + "-" + date[6:8]
```
datetime 모듈 사용해서 format 수정하기

hanseongd commented on scheduler/schedule/api.py:
```diff
@@ -0,0 +1,59 @@
+
+
+def save_holiday(name, is_holiday, date):
+    if not Holiday.objects.filter(date=date, name=name):
```
ORM exist() 함수 사용하기

cafitac replied:
exists() 함수를 사용하는 것이 코딩 패턴에 좋은건가요??

hanseongd replied:
코딩 패턴이 아니라 가져온 필터를 가지고 사용하는게 아닌 존재유무만을 판단하는 경우라서 exist를 써야돼

hanseongd commented on scheduler/schedule/api.py:
```diff
@@ -0,0 +1,45 @@
+        for year in range(2015, 2021):
+            year = str(year)
+            for month in range(1, 13):
+                if month // 10 < 1:
```
if문 필요없음

hanseongd commented on scheduler/schedule/api.py:
```diff
@@ -0,0 +1,45 @@
+        if not Holiday.objects.filter(date=date, name=name).exists():
+            Holiday.objects.create(name=name, is_holiday=is_holiday, date=date)
+
+    def get_api(self):
```
self.holiday_base_url 를 굳이 넣어줄 필요가 있을까?

hanseongd commented on scheduler/schedule/api.py:
```diff
@@ -0,0 +1,45 @@
+                    date = datetime.datetime.strptime(date, "%Y%m%d")
+                    self.save_holiday(name, is_holiday, date)
+
+    def save_holiday(self, name, is_holiday, date):
```
for 문으로 filter를 계속 부르면 db 과부화 일어남
filter는 가능한 적게, 한 번만 쓸 수 있도록 수정해야함

hanseongd replied:
이미 등록된 휴일이 있다면 ?

hanseongd commented on scheduler/schedule/api.py:
```diff
@@ -0,0 +1,46 @@
+class HolidayAPI():
+    def __init__(self):
+        self.api_key = settings.API_KEY # 특일 정보 api 키
+        self.holiday_base_url = "http://apis.data.go.kr/B090041/openapi/service/SpcdeInfoService/{holiday_kinds}?solYear={year}&solMonth={month}&ServiceKey={key}"
```
api key 는 이미 불러온 상태니까 처음부터 적용된 상태로 변경하기

hanseongd commented on scheduler/schedule/api.py:
```diff
@@ -0,0 +1,46 @@
+        self.api_parser("getSundryDayInfo")
+
+
+h = HolidayAPI()
```
python  class의 __call__ func 알아보고 수정해보기

hanseongd commented on scheduler/schedule/api.py:
```diff
@@ -0,0 +1,45 @@
+        self.holiday_base_url = "http://apis.data.go.kr/B090041/openapi/service/SpcdeInfoService/{holiday_kinds}?solYear={year}&solMonth={month}&ServiceKey={key}"
+
+    def __call__(self, *args, **kwargs):
+        self.api_parser("getHoliDeInfo")
```
init에 parse category로 배열 생성하고 for문 돌리기

hanseongd commented on scheduler/schedule/api.py:
```diff
@@ -0,0 +1,45 @@
+
+
+@HolidayAPI
+def get_api():
```
잘못된 형식은 아닌데 데코레이터로 사용하기 보다는 인스턴스 생성하고 실행시키는 방법으로 만들어보자

hanseongd commented on scheduler/schedule/api.py:
```diff
@@ -0,0 +1,43 @@
+        Holiday.objects.get_or_create(name=name, is_holiday=is_holiday, date=date)
+
+
+h = HolidayAPI()
```
h = HolidayAPI()()

hanseongd commented on scheduler/schedule/api.py:
```diff
@@ -0,0 +1,43 @@
+
+class HolidayAPI():
+    def __init__(self):
+        self.holiday_base_url = "http://apis.data.go.kr/B090041/openapi/service/SpcdeInfoService/{{holiday_kinds}}?solYear={{year}}&solMonth={{month}}&ServiceKey={key}".format(key=settings.API_KEY)
```
잘 변경했는데 이런 방법도 있어
self.holiday_base_url = "http://apis.data.go.kr/B090041/openapi/service/SpcdeInfoService/{holiday_kinds}?solYear={year}&solMonth={month}&ServiceKey=%s" % settings.API_KEY

hanseongd commented on scheduler/schedule/models.py:
```diff
@@ -16,4 +16,13 @@ class Schedule(models.Model):
+    date = models.DateField()
+
+    def __str__(self):
+        return "{} [{}]" .format(self.name, self.date)
```
띄어쓰기

Commit:
Holiday api 파싱 스크립트

ADD: api.py로 api 받아오는 스크립트 추가
ADD: Holiday 모델 추가
ADD: Holiday 모델 admin에 등록
MOD: local_settings의 API_KEY 값 변경
